sbi/diagnostics/sbc.py

```diff
@@ -99,60 +89,106 @@ def run_sbc(
         use_batched_sampling=use_batched_sampling,
     )
 
-    # take a random draw from each posterior to get data averaged posterior samples.
+    # Take a random draw from each posterior to get data-averaged posterior samples
     dap_samples = posterior_samples[0, :, :]
-    assert dap_samples.shape == (num_sbc_samples, thetas.shape[1]), "Wrong dap shape."
+    assert dap_samples.shape == (num_sbc_samples, thetas.shape[1]), "Wrong DAP shape."
 
-    ranks = _run_sbc(
-        thetas, xs, posterior_samples, posterior, reduce_fns, show_progress_bar
-    )
+    # Check if we're using log_prob as reduce_fn and if posterior is VIPosterior
+    is_vipost = isinstance(posterior, VIPosterior)
+
+    # Create wrapper for reduce_fns if using a VIPosterior
+    if is_vipost:
+        # Create a wrapper function that ensures the VIPosterior is trained
+        # before applying the reduce function
+        assert isinstance(posterior, VIPosterior), (
+            "VIPosterior is required for this functionality."
+        )
+
+        def make_vipost_wrapper(original_reduce_fn):
+            def wrapped_reduce_fn(theta, x):
+                # Train the posterior on the current x if needed
+                posterior.set_default_x(x)
+                posterior.train(show_progress_bar=False)
+                return original_reduce_fn(theta, x)
+
+            return wrapped_reduce_fn
+
+        # Apply wrapper if reduce_fns is a single callable
+        if callable(reduce_fns) and not isinstance(reduce_fns, str):
+            reduce_fns = make_vipost_wrapper(reduce_fns)
+        # Apply wrapper if reduce_fns is a list of callables
+        elif isinstance(reduce_fns, list):
+            reduce_fns = [make_vipost_wrapper(fn) for fn in reduce_fns]
+
+    # Calculate ranks
+    ranks = _run_sbc(thetas, xs, posterior_samples, reduce_fns, show_progress_bar)
 
     return ranks, dap_samples
 
 
+def _validate_sbc_inputs(
+    thetas: Tensor, xs: Tensor, num_sbc_samples: int, num_posterior_samples: int
+) -> None:
+    """Validate inputs for the SBC procedure."""
+    if num_sbc_samples < 100:
+        warnings.warn(
+            "Number of SBC samples should be on the order of 100s to give reliable "
+            "results.",
+            stacklevel=2,
+        )
+
+    if num_posterior_samples < 100:
+        warnings.warn(
+            "Number of posterior samples for ranking should be on the order "
+            "of 100s to give reliable SBC results.",
+            stacklevel=2,
+        )
+
+    if thetas.shape[0] != xs.shape[0]:
+        raise ValueError("Unequal number of parameters and observations.")
+
+
 def _run_sbc(
     thetas: Tensor,
     xs: Tensor,
     posterior_samples: Tensor,
-    posterior: NeuralPosterior,
```

Comment:
we used this argument only to check whether it's a `VIPosterior`. we now do this above in `run_sbc` so that this internal function can be called with just samples.